security/pkg/nodeagent/caclient/providers/spire/client_test.go

```diff
@@ -0,0 +1,684 @@
+	if sc.latestBundle != newCAPem {
+		t.Errorf("latest bundle was not updated")
+	}
+}
+
+// mockSpireNodeServer is a mock SPIRE server to validate Attestation
+type mockSpireNodeServer struct {
+	// SVIDS to be returned in attestation response
+	Svids [][]byte
+	// CAs bundles to be returned in attestation response
+	CAs [][]byte
+	// error to be returned in case it is expected
+	Err error
+	// spiffeID in attestation response map
+	CertSpiffeID string
+	// trust domain in bundles map
+	TrustDomain string
+}
+
+// Attest mock attestation logic to verify attestation requests
+func (s mockSpireNodeServer) Attest(stream spireIntegration.Node_AttestServer) error {
+	if s.CertSpiffeID == "" {
+		s.CertSpiffeID = defaultCertificateSpiffeID
+	}
+	if s.TrustDomain == "" {
+		s.TrustDomain = defaultTrustDomain
+	}
+
+	// return error in case it is expected
+	if s.Err != nil {
+		return s.Err
+	}
+
+	req, err := stream.Recv()
+	if err != nil {
+		return fmt.Errorf("could not recover attest stream: %v", err)
+	}
+
+	if req.AttestationData.Type != "istio" {
+		return errors.New("unexpected attestor type")
+	}
+
+	var data IstioAttestableData
+	err = json.Unmarshal(req.AttestationData.Data, &data)
+	if err != nil {
+		return fmt.Errorf("could not unmarshal istio attested data: %v", err)
+	}
+
+	svids := make(map[string]*spireIntegration.X509SVID)
+	for _, svid := range s.Svids {
+
+		svids[s.CertSpiffeID] = &spireIntegration.X509SVID{
+			CertChain: svid,
+		}
+	}
+
+	bundles := make(map[string]*spireIntegration.Bundle)
+	if len(s.CAs) > 0 {
+		var roots []*spireIntegration.Certificate
+		for _, ca := range s.CAs {
+			roots = append(roots, &spireIntegration.Certificate{DerBytes: ca})
+		}
+
+		bundles[s.TrustDomain] = &spireIntegration.Bundle{
+			RootCas: roots,
+		}
+	}
+
+	resp := &spireIntegration.AttestResponse{
+		SvidUpdate: &spireIntegration.X509SVIDUpdate{
+			Svids:   svids,
+			Bundles: bundles,
+		},
+	}
+
+	stream.Send(resp)
+	return nil
+}
+
+// createTestServer creates a server with node attestor for testing purposes
+func createTestServer(t *testing.T, bootstrapCert *x509.Certificate, cert *x509.Certificate, key crypto.PrivateKey,
+	server *mockSpireNodeServer, id string) (*grpc.Server, string) {
+	var tlsCertificate [][]byte
+	tlsCertificate = append(tlsCertificate, cert.Raw)
+	certPool := x509.NewCertPool()
+	certPool.AddCert(bootstrapCert)
+
+	lis, err := net.Listen("tcp", mockServerAddress)
+	if err != nil {
+		t.Fatalf("Test case [%s]: failed to listen: %v", id, err)
+	}
+
+	s := grpc.NewServer(
+		grpc.Creds(credentials.NewTLS(&tls.Config{
+			Certificates: []tls.Certificate{
+				{
+					Certificate: tlsCertificate,
+					PrivateKey:  key,
+				},
+			},
+			ClientCAs:  certPool,
+			ClientAuth: tls.VerifyClientCertIfGiven,
+		})))
+
+	spireIntegration.RegisterNodeServer(s, server)
+	go s.Serve(lis)
+
+	return s, lis.Addr().String()
+}
+
+// writeKey takes a private key, formats as PEM, and writes it to a decodeKey
```

Comment:
The function name "writeKey" in the comment is different from the actual function name "decodeKey".